File: reverse_binary_multiplication.c
#include <stdio.h>
#include <stdlib.h>

unsigned long long revere_binary_multiplication( unsigned long long multiplier, unsigned long long target_product, int bit_size){

    unsigned long long current_product_sum = 0;
    unsigned long long missing_factor = 0;
    
    int binary_ptr = 0;
    while (!((multiplier >> binary_ptr) & 1ULL) && binary_ptr < bit_size) {
        if (((target_product >> binary_ptr) & 1ULL) == 1ULL) {
            return 0;
        }
        binary_ptr++;
    }
    if (binary_ptr == bit_size && target_product != 0) {
        return 0;
    }
    for (int factor_bit_ptr = 0; (factor_bit_ptr + binary_ptr) < bit_size; ++factor_bit_ptr) {
        
        int target_bit = (target_product >> (factor_bit_ptr + binary_ptr)) & 1ULL;
        int product_bit = (current_product_sum >> (factor_bit_ptr + binary_ptr)) & 1ULL;
        if (target_bit != product_bit) {
            missing_factor |= (1ULL << factor_bit_ptr);
            current_product_sum += (multiplier << factor_bit_ptr );
        }
    }
    return missing_factor;
}

int main(){
    unsigned long long a, b, n, x;
    printf("Reverse Binary Multiplication\n");
    printf("to solve: a*x = b (mod 2^n)\n");

    printf("Enter a: ");
    scanf("%llu",&a);
    printf("Enter b: ");
    scanf("%llu",&b);
    printf("Enter n: ");
    scanf("%llu",&n);

    x = revere_binary_multiplication(a, b, n);
    if (x==0)
        printf("Solution Doesn't Exist\n");
    else
        printf("solution x = %llu\n",x);

    return 0;
}

File: reverse_binary_multiplication.py
# reverse the binary multiplication to find the factor under under modulo 2^n
# this function find x if : a * x = b mod 2^n  [ where a, b and n are given ]
def reverse_binary_multiplication(multiplier, target_product, bit_size):
    # make a buffer to the size of the array to bit_size for calculating the product
    binary_buffer = 0
    missing_factor = 0

    # setting the pointer where the first 1 appeared in the binary_multiplier from Least Significant Bit
    binary_ptr = 0
    while ((multiplier >> binary_ptr) & 1) != 1:
        if ((target_product >> binary_ptr) & 1) == 1:
            return
        binary_ptr +=1

    
    factor_ptr = 0
    target_ptr = binary_ptr
    while target_ptr < bit_size:
        buffer_index = bit_size - target_ptr -1 
        if  binary_buffer>>target_ptr & 1 != (target_product>>target_ptr) & 1: 
            missing_factor += 1<<factor_ptr
            binary_buffer += multiplier << factor_ptr
            target_ptr += 1
            factor_ptr+=1
        else:
            target_ptr += 1
            factor_ptr+=1
    
    return missing_factor


if __name__=='__main__':
    import time
    print("Reverse Binary Multiplicaiton Method")
    print("solve 'x' if : a * x = b mod 2^n [for given a and b under modulo 2^n]")
    print()
    multiplier = int(input("enter multiplier(a) : "))
    target_product = int(input("enter target product(b) : "))
    bit_size = int(input("enter bit size (n) : "))
    time_start = time.time()
    missing_factor = reverse_binary_multiplication(multiplier, target_product, bit_size)
    if missing_factor == None:
        print("solution dosn't exist !!!")
        print("number-of-zeros-in-LSB-in-target_product > number-of-zeros-in-LSB-in-muliplier")
        quit()
    print(f" missing number is {missing_factor}")
    print(f" impossible equation solved: {missing_factor} * {multiplier} = {target_product}")
    print(f"found in {time.time()-time_start}sec..")